Select Group implementation + Fixes

akshatsinghkaushik opened the pull request:
Added features:
- User can select from a list of groups they are a part of
- User can switch between group chats they are a part of
- Users registering on the App using Google or Github now have a default display name obtained from Google/Github
-- Minor bug fixes implemented

Jahmilli commented:
when i try and create a group settign the leaders and partcipants as sebastian.southern@gmail.com as well as apple@apple.com nothing seems to happen?

akshatsinghkaushik commented:
Fixed: 
![image](https://user-images.githubusercontent.com/37283605/82754620-7010c000-9e11-11ea-9793-c8a29ef03435.png)
Make sure you check the box next to the name before clicking save.
![image](https://user-images.githubusercontent.com/37283605/82754632-9171ac00-9e11-11ea-81f6-d21e71a1aa58.png)

![image](https://user-images.githubusercontent.com/37283605/82754676-e7465400-9e11-11ea-8bf7-557e3fd7feec.png)
![image](https://user-images.githubusercontent.com/37283605/82754686-f75e3380-9e11-11ea-98e8-2295e15d9881.png)

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -48,67 +51,103 @@ class Chat extends Component {
+          }
         });
+        this.setState({ groups: groups_temp });
+        if (this.state.selectedGroupID === "") {
```
if you're checking for empty strings its cleaner just to do `if (!this.state.selectedGroupID)`

akshatsinghkaushik replied:
Fixed, Thanks!

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -48,67 +51,103 @@ class Chat extends Component {
+      this.setState({ readError: error.message });
     }
+
+    let groups_list = new Map();
```
everything else is 🐫cased and now we got 🐍case
Also i think this should be const

akshatsinghkaushik replied:
Fixed, Thanks!

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -48,67 +51,103 @@ class Chat extends Component {
+        db.ref(`groups/${this.state.selectedGroupID}/chats`).once(
+          "value",
+          (snapshot) => {
+            let chats = [];
```
'chats' doesnt get reassigned so this should be `const`

akshatsinghkaushik replied:
Fixed, Thanks!

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -48,67 +51,103 @@ class Chat extends Component {
+              chats.push(snap.val());
+            });
+
+            chats.sort(function (a, b) {
```
would be slightly nicer to use an arrow function here

akshatsinghkaushik replied:
Fixed, Thanks!

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -117,10 +156,61 @@ class Chat extends Component {
+        selectedGroupName: this.state.groups.get(event.target.id).name,
+      });
+
+      db.ref(`groups/${event.target.id}/chats`).on("value", (snapshot) => {
```
im a little bit unsure why we create two listeners, one in componentDidMount and one here

akshatsinghkaushik replied:
Fixed, Thanks!
Replaced with a single call to update the entries instead of opening an event listener.

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -117,10 +156,61 @@ class Chat extends Component {
   }
 
+  handleSelectGroup = (event) => {
+    if (event.target.id !== "") {
```
just do !event.target.id

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -117,10 +156,61 @@ class Chat extends Component {
+        this.setState({ chats });
+      });
+      db.ref(`groups/${this.state.selectedGroupID}/chats`).off("value");
+    } else {
```
In what cases will  `event.target.id` empty that this gets triggered?

akshatsinghkaushik replied:
The conditional statement is to capture Event Bubbling scenarios, 
If (user clicks on the Outer Group Box containing GroupName, description divs) then use event.target.id (Since id of the Group box = GroupUID)
Else use event.currentTarget.id (Since id of parent Box = GroupUID)

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -183,40 +273,42 @@ class Chat extends Component {
+                {Array.from(this.state.groups.values()).map((result, index) => {
+                  return (
+                    <div
+                      key={index}
```
I think its better that we use the groupId as the key here

akshatsinghkaushik replied:
Fixed, Thanks!

Jahmilli commented on src/Components/Pages/Chat.js:
```diff
@@ -236,7 +328,11 @@ class Chat extends Component {
+                    <strong className="text-info">
+                      {this.state.selectedGroupName
+                        ? this.state.selectedGroupName
+                        : "Group Name"}
```
instead of a ternary operator here you can do `{this.state.selectedGroupName || "Group Name"}`
if selectedGroupName is empty than it should return false and "Group Name" would be displayed

akshatsinghkaushik replied:
Fixed, Thanks!

Jahmilli commented on src/Components/Pages/Modal.js:
```diff
@@ -153,6 +153,26 @@ export default function AnimatedModal(props) {
+  useEffect(() => {
+    try {
+      db.ref("groups").on("value", (snapshot) => {
+        let groups_temp = new Map();
```
const

akshatsinghkaushik replied:
Fixed, Thanks!

Jahmilli commented on src/helpers/auth.js:
```diff
@@ -50,8 +51,9 @@ export function signInWithGitHub() {
-      await db.ref("users/" + user.uid).set({
+      await db.ref("users/" + user.uid).update({
         email: user.email,
+        name: user.displayName,
```
I think we should keep all these fb update/get/create db functions in db.js so we can reuse them but also ive been doing my changes in a fairly unstructured manner so it might be messy...

Jahmilli commented:
chat seems to be broken atm :(

akshatsinghkaushik commented:
> chat seems to be broken atm :(
> Can send messages, but they dont appear immediately.
> If i refresh i can see messages

Fixed, brought back an event listener which I had removed earlier which fixing something else.
